rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,676 @@
+   --lang-warn-recursive=6
+   --lang-warn-verbose
+   ```
+
+   The naming may need some bikeshedding. For example, one could use the same syntax as with C-compilers (probably not though):
+
+   - `-Wall`
+   - `-Wno-`
+   - `-Werror=`
+
+   <details><summary>Arguments</summary>
+
+   - (+) Disallowing reuse avoids dealing with change of meaning across versions
+   - (+) Not considering experimental features simplifies their handling and is not required by them being exempt from compatibility guarantees.
+   - (-) Maintenance burden for everyone using these old constructs, or evaluating old revisions of Nixpkgs.
+   
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Opt-in warnings
+     - (-) Can't really make breaking changes as people won't be warned ahead of time
+   - No warnings, just errors
+     - (-) Doesn't offer a transition window to users
+     - (+) Easier to implement
+
+   </details>
+
+1. A the end of the evaluation, print statistics and explanations.
+   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Show the current Nix language version
+
+```console
+nix --language-version
+7
+```
+
+```console
+nix --supported-language-versions
+6
+7
+```
+
+## Version interoperability
+
+### Versioned built-ins can be passed across file boundaries
+
+```nix
+# a.nix
+version 6;
+builtins.langVersion
+```
+
+```nix
+# b.nix
+version 7;
+[ (import ./a.6.nix) builtins.langVersion ]
+```
+
+```console
+$ nix-instantiate --eval b.nix
+[ 6 7 ]
+```
+
+### Expressions are not forward compatible
+
+```nix
+# a.nix
+version 6;
+import ./b.nix
+```
+
+```nix
+# b.nix
+version 7;
+builtins.langVersion
+```
+
+```console
+$ nix-instantiate --eval a.nix
+error: unsupported Nix language version 7
+```
+
+
+### Best-effort interoperability
+
+```nix
+# a.nix
+version 6;
+{ increment }:
+increment 1
+```
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { increment = x: x + 1 }
+```
+
+Since `increment` written in version 7 carries its own implementation with it, forcing it within an expression written in version 6 just works:
+
+```console
+$ nix-instantiate --eval b.nix
+2
+```
+
+### Pathological example
+
+Usually existing code will be interacted with by calling functions.
+When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.
+
+```nix
+# a.nix
+version 6;
+{ value }:
+value + 1
+```
+
+Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { value = %5 + 7i%; }
+```
+
+```console
+$ nix-instantiate --eval bnix
+error: unsupported value type `complex` at built-in operator `+`
+```
+
```

Comment:
````suggestion

In the following example, version 7 *removed* the `null` primitive, such that it can no longer be used.

```nix
# a.nix
null
```

```nix
# b.nix
version 7;
import ./a.nix
```

```console
$ nix-instantiate --eval b.nix
error: unsupported value `null`
```
````

This is consistent with best-effort interoperability:
Old code keeps working on its own, and new code has to be adapted because it was not there before the breaking change.

While this requires additional effort to adopt the new language version, expression authors can always recourse to writing new code in older versions while using newer evaluators.
This in fact allows for creating compatibility wrappers as needed.

```